docs/articles/documentation/recipes/best-practices/best-practices.md

```diff
@@ -0,0 +1,223 @@
+---
+layout: docs
+title: Best Practices
+permalink: /documentation/recipes/best-practices/best-practices.html
+---
+# Best Practices
+
+This article covers the recommended ways to test with TestCafe and covers the following topics:
+
+* Smart Assertions
+* Use of Page Objects
+* Use of Roles for Login
+* File structure
+* Clean up between Tests (beforeEach)
+* E2E Tests Limitations
+* Selectors strategy
+
+## Smart Assertions
+
+In end-to-end web testing, peripheral factors (like network lag, processor, or memory bottlenecks in containers) can interfere with the assertions and render tests inconclusive (otherwise known as 'flaky').
+
+Out of the box, TestCafe includes a [Smart Assertion Query Mechanism](../../guides/basic-guides/assert.md#smart-assertion-query-mechanism), which helps mitigate these effects. This mechanism introduces additional wait time for all the assertions: if an assertion fails, the test doesn't fail straight away. Instead, the assertion retries multiple times within a timeout. That helps to stabilize the tests without a big impact on performance.
+
+However, the smart assertions query mechanism only works with promises returned from TestCafe's own `Selector API`.
+
+The following example demonstrates the potential issue:
+
+```js
+import { Selector } from 'testcafe';
+
+fixture `My fixture`
+    .page `http://devexpress.github.io/testcafe/example/`;
+
+test('Assertion with Selector', async t => {
+    const developerNameInput = Selector('#developer-name');
+
+    //this technique leads to unstable results. Avoid it.
+    const developerName      = await Selector('#developer-name').value;
+    //
+
+    console.log(developerName);
+
+    await t
+        .wait(1000)
+        .expect(developerNameInput.value).eql('')
+        .typeText(developerNameInput, 'Peter')
+        .expect(developerName).eql('Peter');
```

Comment:
Lines 41 and 45 confused even me. We output a blank string, then check that the input value is blank...

Just a suggestion, in case you think this is more illustrative:

```js
    await t.typeText(developerNameInput, 'Peter');

    const developerName = await Selector('#developer-name').value;

    await t
         .expect(developerName).eql('Peter')
         .typeText(developerNameInput, 'Jack')
         .expect(developerName).eql('Jack'); // fails
```

Reply:
await t
        .expect(developerNameInput.value).eql('') //We check that the input is empty on page load
        .typeText(developerNameInput, 'Peter') //We type the text
        .expect(developerName).eql('Peter'); //We check the value

Let me know if that doesn't work the way I think it does and I'll use your sample.